crates/wasm-encoder/src/code.rs

```diff
@@ -0,0 +1,831 @@
+use super::*;
+
+/// An encoder for the code section.
+///
+/// # Example
+///
+/// ```
+/// use wasm_encoder::{
+///     CodeSection, Function, FunctionSection, Instruction, Module,
+///     TypeSection, ValType
+/// };
+///
+/// let mut types = TypeSection::new();
+/// types.function(vec![], vec![ValType::I32]);
+///
+/// let mut functions = FunctionSection::new();
+/// let type_index = 0;
+/// functions.function(type_index);
+///
+/// let locals = vec![];
+/// let mut func = Function::new(locals);
+/// func.instruction(Instruction::I32Const(42));
+/// let mut code = CodeSection::new();
+/// code.function(&func);
+///
+/// let mut module = Module::new();
+/// module
+///     .section(&types)
+///     .section(&functions)
+///     .section(&code);
+///
+/// let wasm_bytes = module.finish();
+/// ```
+pub struct CodeSection {
+    bytes: Vec<u8>,
+    num_added: u32,
+}
+
+impl CodeSection {
+    /// Create a new code section encoder.
+    pub fn new() -> CodeSection {
+        CodeSection {
+            bytes: vec![],
+            num_added: 0,
+        }
+    }
+
+    /// Write a function body into this code section.
+    pub fn function(&mut self, func: &Function) -> &mut Self {
+        func.encode(&mut self.bytes);
+        self.num_added += 1;
+        self
+    }
+}
+
+impl Section for CodeSection {
+    fn id(&self) -> u8 {
+        SectionId::Code as u8
+    }
+
+    fn encode<S>(&self, sink: &mut S)
+    where
+        S: Extend<u8>,
```

Comment:
Out of curiosity does this optimize well? I'd imagine that while this is super general it'll rarely get used with much other than `Vec`

Reply:
I just did it because it gives less capabilities to implementations. I originally had a `&mut Vec<u8>` and a comment like "only push data, don't pop anything" and then I realized I could just make it an `Extend<u8>` to enforce that at the type level.